plugins/preformatted.py

```diff
@@ -0,0 +1,169 @@
+from typing import Dict, Union
+
+import bs4
+import pygments as pyg
+from pygments import lexers, util
+
+from util import LINE_SEPARATOR
+from plugins.syntax_highlighting import IndFormatter
+
+Options = Dict[str, Union[str, bool]]
+
+# Maximum length of a line in a code block
+MAX_PRE_LINE_LENGTH = 48
+
+
+def highlight_code(pre_contents: str, options: Options) -> str:
+    """ Add syntax highlighting to a code block
+    """
+    # Find lexer for the given language
+    lang_name = options.get('language', options.get('lang', None))  # allow for language or lang options
+    try:
+        lexer = pyg.lexers.get_lexer_by_name(lang_name)
+    except pyg.util.ClassNotFound:
+        return pre_contents
+
+    # Highlight the code
+    formatter = IndFormatter(style='bw')
+    pre_text = bs4.BeautifulSoup(pre_contents, 'html.parser').get_text().strip()
+    pre_text = pyg.highlight(pre_text, lexer, formatter)
+    # Strip ending whitespace
+    return pre_text.rstrip()
+
+
+def add_linenos(pre_contents: str, options: Options) -> str:
+    """ Add line numbers to a code block
+    """
+    if options.get('linenos', False):
+        first_line, *rest = pre_contents.split('\n')
+        pre_contents = f'<mathnews-pre--lineno-start>{first_line}</mathnews-pre--lineno-start>'
+        if rest:
+            pre_contents += '\n<mathnews-pre--lineno>{}</mathnews-pre--lineno>'.format('\n'.join(rest))
```

Comment:
nit: this line is kinda long and confusing, recommend pulling the joined line into a separate variable `rest_lines`, and using f-strings

Reply:
Named it `rest_joined`, hopefully that doesn't ruin things.